src/lib/mdns/tests/TestTxtFields.cpp

```diff
@@ -0,0 +1,377 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <mdns/TxtFields.h>
+
+#include <limits>
+#include <string.h>
+
+#include <mdns/Resolver.h>
+#include <support/UnitTestRegistration.h>
+
+#include <nlunit-test.h>
+
+using namespace chip;
+using namespace chip::Mdns;
+using namespace chip::Mdns::Internal;
+
+namespace {
+
+ByteSpan GetSpan(char * key)
+{
+    return ByteSpan(reinterpret_cast<uint8_t *>(key), strlen(key));
+}
+
+void TestGetTxtFieldKey(nlTestSuite * inSuite, void * inContext)
+{
+    char key[3];
+    sprintf(key, "D");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kLongDiscriminator);
+
+    sprintf(key, "VP");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kVendorProduct);
+
+    sprintf(key, "AP");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kAdditionalPairing);
+
+    sprintf(key, "CM");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kCommissioningMode);
+
+    sprintf(key, "DT");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kDeviceType);
+
+    sprintf(key, "DN");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kDeviceName);
+
+    sprintf(key, "RI");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kRotatingDeviceId);
+
+    sprintf(key, "PI");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kPairingInstruction);
+
+    sprintf(key, "PH");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kPairingHint);
+
+    sprintf(key, "XX");
+    NL_TEST_ASSERT(inSuite, GetTxtFieldKey(GetSpan(key)) == TxtFieldKey::kUnknown);
+}
+void TestGetProduct(nlTestSuite * inSuite, void * inContext)
+{
+    // Product and vendor are given as part of the same key, on either side of a + sign. Product is after the +
+    char vp[64];
+    sprintf(vp, "123+456");
```

Comment:
This does not test the fact that the incoming spans will not be nul-terminated.

Test code should enforce bad data past the span.

Suggest the following pattern:

```
sprintf(vp, "123+");
ByteSpan vp_span = GetSpan(vp);
// Force underlying buffer to have a digit rather than a null terminator
// to validate that span is not assumed to have nul-terminated backing.
vp[strlen(vp)] = '1';
NL_TEST_ASSERT(inSuite, GetProduct(vp_span) == 0);
```

Or something similar done through GetSpan, so that the code being tested can validate to some degree that the bounds of the span are not crossed and interpreted.